In [1]:
#Important Libraries
import pandas as pd

The researchers were not certain whether the description is an important feature or not because it might contain redundant information and it is making the size of the dataset huge. Also, tha data set is composed of 31 files because of its huge size. As a result, we decided to upload only one file and to decide whether to include the description or not. And based on this, we will do the same with the remaining files.


In [2]:
df = pd.read_csv("used_cars_data-1.csv")

In [3]:
df.isna().sum()


vin                         0
back_legroom             4954
bed                     99393
bed_height              91877
bed_length              91877
                        ...  
wheel_system             4624
wheel_system_display     4624
wheelbase                4954
width                    4954
year                        0
Length: 66, dtype: int64

In [4]:
df1_random_sample_of_one_thousand_obs= df.sample(n = 100)

In [5]:
df1_random_sample_of_one_thousand_obs.to_csv("sample_of_one_thousand_observation.csv")

Kindly open the downloaded csv file named sample_of_one_thousand_observation to have a better understanding of the decription column in relation to other variables.

The reserachers found that the decription variable carries no additional information about the car. In addition, removing the column Description decreases the file size by 85% which makes the csv file lighter and easier to deal with. Upon reading the observation, the descritption column does not carry additional info than those represented in the other columns. Thus it will be reomved. Since, the size of the individual file is about 350 Megabytes. As a result, it would be hard to upload all of them. So, we decided to remove the description column using excel from the files and then we will upload and concatenate them to complete preprocessing

In [6]:
df1 = pd.read_csv("/content/drive/MyDrive/ML Data/used_cars_data-1.csv")
for i in range(2, 32):
  path = "/content/drive/MyDrive/ML Data/used_cars_data-" + str(i)+ ".csv"
  dft = pd.read_csv(path)
  df1 = df1.append(dft)

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (11) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)
/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


We loaded all the files without the description into a single pandas dataframe.

At first, we want to investigate the different body types of the cars we have.

In [7]:
df1["body_type"].value_counts()

SUV / Crossover    1416402
Sedan               742036
Pickup Truck        474595
Hatchback            88374
Minivan              79802
Coupe                71607
Van                  47166
Wagon                40505
Convertible          26010
Name: body_type, dtype: int64

In [8]:
dft = df1[df1.body_type == "Pickup Truck"]
dft = dft.append(df1[df1.body_type == "Van"])
print("Percentage of Pickup trucks and Vans = ", len(dft)/len(df1)*100)
print("Percentage pf null values of bed feature = ", df1.bed.isna().sum()/ len(df1)*100)
print("Percentage pf null values of bed length feature = ", df1.bed_length.isna().sum()/ len(df1)*100)
print("Percentage pf null values of bed height feature = ", df1.bed_height.isna().sum()/ len(df1)*100)
print("Unique values of bed length feature = ", df1.bed_length.unique())
print("Unique values of bed height feature = ", df1.bed_height.unique())
print("Percentage of Null values of body type feature = ", df1['body_type'].isna().sum()/len(df1)*100)

Percentage of Pickup trucks and Vans =  17.39180144264743
Percentage pf null values of bed feature =  99.34774203010626
Percentage pf null values of bed length feature =  85.69692404101279
Percentage pf null values of bed height feature =  85.69692404101279
Unique values of bed length feature =  [nan '78.9 in' '--' '81.9 in' '74 in' '67.4 in' '69.9 in' '79.4 in'
 '69.3 in' '61.7 in' '60.5 in' '67.1 in' '78.8 in' '76.3 in' '78.7 in'
 '60 in' '98.3 in' '72.7 in' '60.3 in' '71.7 in' '98.2 in' '97.6 in'
 '98 in' '98.6 in' '73.5 in' '98.1 in' '81.8 in' '66.7 in' '59.5 in'
 '63.3 in' '77.4 in' '82.2 in' '73.3 in' '73.7 in' '72.8 in' '72.3 in'
 '82.3 in' '61 in' '97.8 in' '61.1 in' '76.5 in' '82.4 in' '63 in'
 '67.3 in' '64.9 in' '74.3 in' '97.4 in' '67 in' '55.2 in' '50 in'
 '69.2 in' '74.5 in' '71.8 in' '67.2 in' '79.1 in' '64 in' '98.5 in'
 '56.3 in' '49.6 in' '41.5 in' '78 in' '66 in' '77.5 in' '59.3 in'
 '63.7 in' '75.9 in' '78.6 in' '83.8 in' '74.6 in' '76.8 in' '97 in'
 '96 in' '98.8 i

From the previous values, we can see that cars of body types: Pickup Trucks and Vans constitute only 17.4 percent of the data. Also, the features associated with them such as the bed, bed_length, bed_height is mostly composed of null values and the non values are not indicative such as in bed height as data are either nulls or "--". Bed length is better but it also contains sum values which are "--". So, we decided to remove the pickup trucks and the vans and also remove the features: bed, bed_length, and bed_height. Also, we decided to remove the rows having null values of in the body type feature since their percentage is low when compared to the total number of samples.

In [9]:
df1 = df1[df1.body_type != "Pickup Truck"]
df1= df1[df1.body_type != "Van"]
df1= df1[df1[['body_type']].notnull().all(1)]
df1 = df1.drop("bed", axis = 1)
df1 = df1.drop("bed_height", axis = 1)
df1 = df1.drop("bed_length", axis = 1)

For the feature of the photo URL of the car, we decided that we will not do any kind of image processing. As a result, this feature will not be of importance of us. So, we decided to remove it.

In [10]:
df1 = df1.drop("main_picture_url", axis = 1)

In [11]:
print("Percentage of Null values of combined fuel economy feature = ", df1['combine_fuel_economy'].isna().sum()/len(df1)*100)
print("Percentage of Null values of is_certified feature = ", df1['is_certified'].isna().sum()/len(df1)*100)

Percentage of Null values of combined fuel economy feature =  100.0
Percentage of Null values of is_certified feature =  100.0


Since all of the values of combined fuel economy feature and the is certified feature are nulls, we decided to remove those features.

In [12]:
df1 = df1.drop("is_certified", axis = 1)
df1 = df1.drop("combine_fuel_economy", axis = 1)

In [15]:
print("Percentage of Null values of cabin feature = ", df1['cabin'].isna().sum()/len(df1)*100)
print("Percentage of Null values of is_oemcpo feature = ", df1['is_oemcpo'].isna().sum()/len(df1)*100)

Percentage of Null values of cabin feature =  99.99833653583995
Percentage of Null values of is_oemcpo feature =  95.12678031237422


Since a huge percentage of the cabin - a feature for trucks and vans-   feature and the is_oemcpo - a feature describing whether Pre-owned cars were certified by the manufacturer-  feature are nulls, we decided to remove those features.

In [16]:
df1 = df1.drop("cabin", axis = 1)
df1 = df1.drop("is_oemcpo", axis = 1)

In [18]:
print("Percentage of Null values of vehicle_damage_category feature = ", df1['vehicle_damage_category'].isna().sum()/len(df1)*100)


Percentage of Null values of vehicle_damage_category feature =  100.0


Since all values of the vehicle damage category feature are nulls, we decided to remove this feature.

In [19]:
df1 = df1.drop("vehicle_damage_category", axis = 1)

In [23]:
print("Percentage of Null values of is cpo feature = ", df1['is_cpo'].isna().sum()/len(df1)*100)


Percentage of Null values of is cpo feature =  93.19996137517364


Since most of the values of the is cpo feature are nulls, we decided to remove this feature

In [24]:
df1 = df1.drop("is_cpo", axis = 1)

Now, we will get a random sample from the data. We agreed to take 200000 samples.

In [25]:
Data = df1.sample(n = 200000)

In [26]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 57071 to 83398
Data columns (total 55 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   vin                   200000 non-null  object 
 1   back_legroom          192540 non-null  object 
 2   body_type             200000 non-null  object 
 3   city                  200000 non-null  object 
 4   city_fuel_economy     176380 non-null  float64
 5   daysonmarket          200000 non-null  int64  
 6   dealer_zip            200000 non-null  object 
 7   engine_cylinders      194920 non-null  object 
 8   engine_displacement   191357 non-null  float64
 9   engine_type           194920 non-null  object 
 10  exterior_color        199996 non-null  object 
 11  fleet                 108332 non-null  object 
 12  frame_damaged         108332 non-null  object 
 13  franchise_dealer      200000 non-null  bool   
 14  franchise_make        160488 non-null  object 
 1

In [27]:
Data.to_csv("Final_Sample.csv")

Now, we saved the sample to a csv file. We will download it and work on it in another Notebook.